sklearn/linear_model/sgd_fast.pyx

```diff
@@ -705,13 +756,35 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
                 infinity = True
                 break
 
-            if tol > -INFINITY and sumloss > previous_loss - tol * n_samples:
-                if verbose:
-                    with gil:
-                        print("Convergence after %d epochs took %.2f seconds"
-                              % (epoch + 1, time() - t_start))
-                break
-            previous_loss = sumloss
+            # evaluate the score on the validation set
+            if early_stopping:
+                with gil:
+                    score = estimator._validation_score(weights, intercept)
+                if tol > -INFINITY and score < previous_score + tol:
+                    no_improvement_count += 1
+                else:
+                    no_improvement_count = 0
+                previous_score = score
```

Comment:
We should not use the `previous_score` / `previous_loss` but the `best_score` and `best_loss` observed so far to increment or reset the `no_improvement_count` counter.

Reply:
The following sequences of validation scores should trigger early stopping with a patience of 5:

0.7, 0.8, 0.9, 0.79, 0.87, 0.81, 0.82

0.9 is the best score and there was no improvement for 5 consecutive iterations thereafter even if there has been many fluctuations.